File: csharp/DesiArchitect.DesignPatterns/05_Singleton/Before/AppConfigStatic.cs
namespace DesiArchitect.DesignPatterns.Singleton.Before;

// Textbook singleton: simple syntax, tight coupling, hard to test
public class AppConfigStatic
{
    private static AppConfigStatic? _instance;

    public static AppConfigStatic Instance => _instance ??= new AppConfigStatic();  // global access point

    public string ApiKey => "sk-live-demo-key-12345";
    public string DatabaseConnection => "Server=prod-db;Database=Orders;";

    private AppConfigStatic()
    {
        Console.WriteLine("  [Before] AppConfigStatic created (static instance)");
    }

    public void ShowConfig()
    {
        Console.WriteLine($"  [Before] AppConfig.Instance accessed - tightly coupled, unmockable");
        Console.WriteLine($"     API Key: {ApiKey}");
        Console.WriteLine($"     DB Conn: {DatabaseConnection}");
    }
}

File: typescript/src/05-singleton/before/app-config-static.ts
export class AppConfigStatic {
  private static _instance: AppConfigStatic | undefined;

  static get instance(): AppConfigStatic {
    if (!AppConfigStatic._instance) {
      AppConfigStatic._instance = new AppConfigStatic();
    }
    return AppConfigStatic._instance;
  }

  readonly apiKey = 'sk-live-demo-key-12345';
  readonly databaseConnection = 'Server=prod-db;Database=Orders;';

  private constructor() {
    console.log('  [Before] AppConfigStatic created (static instance)');
  }

  showConfig(): void {
    console.log('  [Before] AppConfig.Instance accessed - tightly coupled, unmockable');
    console.log(`     API Key: ${this.apiKey}`);
    console.log(`     DB Conn: ${this.databaseConnection}`);
  }
}